backend.go

```diff
@@ -0,0 +1,130 @@
+package jwtauth
+
+import (
+	"context"
+	"crypto/tls"
+	"crypto/x509"
+	"errors"
+	"net/http"
+	"sync"
+	"sync/atomic"
+
+	oidc "github.com/coreos/go-oidc"
+	cleanhttp "github.com/hashicorp/go-cleanhttp"
+	"github.com/hashicorp/vault/logical"
+	"github.com/hashicorp/vault/logical/framework"
+	"golang.org/x/oauth2"
+)
+
+// Factory is used by framework
+func Factory(ctx context.Context, c *logical.BackendConfig) (logical.Backend, error) {
+	b := backend(c)
+	if err := b.Setup(ctx, c); err != nil {
+		return nil, err
+	}
+	return b, nil
+}
+
+type jwtAuthBackend struct {
+	*framework.Backend
+
+	l sync.RWMutex
+
+	provider      *oidc.Provider
+	parsedPubKeys atomic.Value
+}
+
+func backend(c *logical.BackendConfig) *jwtAuthBackend {
+	b := new(jwtAuthBackend)
+
+	b.Backend = &framework.Backend{
+		AuthRenew:   b.pathLoginRenew,
+		BackendType: logical.TypeCredential,
+		Invalidate:  b.invalidate,
+		Help:        backendHelp,
+		PathsSpecial: &logical.Paths{
+			Unauthenticated: []string{
+				"login",
+			},
+			SealWrapStorage: []string{
+				"config",
+			},
+		},
+		Paths: framework.PathAppend(
+			[]*framework.Path{
+				pathLogin(b),
+				pathConfig(b),
+			},
+		),
+	}
+
+	// Seed the type
+	b.parsedPubKeys.Store(([]interface{})(nil))
```

Comment:
Any reason why not to control this with the mutex?  I don't see a high concurrent workload for these values changing and the provider is already protected with the mutex.

Reply:
Done.